retail/retail-java-applications/terraform/test/pubsub_test.go

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package test
+
+import (
+	"cloud.google.com/go/pubsub"
+	"context"
+	"fmt"
+	"github.com/gruntwork-io/terratest/modules/gcp"
+	"github.com/gruntwork-io/terratest/modules/terraform"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestPubsub(t *testing.T) {
+
+	projectID := gcp.GetGoogleProjectIDFromEnvVar(t)
+	opts := &terraform.Options{
+		TerraformDir: "../examples/pubsub",
+		Vars:         map[string]interface{}{"project_id": projectID},
+	}
+	ctx := context.Background()
+	client, err := pubsub.NewClient(ctx, projectID)
+	if err != nil {
+		t.Fatalf("NewClient: %v", err)
+	}
+
+	cleanup(ctx, client, t)
+
+	defer terraform.Destroy(t, opts)
+	terraform.Init(t, opts)
+	terraform.Apply(t, opts)
+
+	testCases := []struct {
+		name      string
+		topicName string
+		subName   string
+		expected  string
+	}{
+		{
+			"ClickstreamInbound",
+			"Clickstream-inbound",
+			"Clickstream-inbound-sub",
+			"Clickstream-inbound-test-message",
+		},
+		{
+			"InventoryInbound",
+			"Inventory-inbound",
+			"Inventory-inbound-sub",
+			"Inventory-inbound-test-message",
+		},
+		{
+			"TransactionsInbound",
+			"Transactions-inbound",
+			"Transactions-inbound-sub",
+			"Transactions-inbound-test-message",
+		},
+	}
+
+	t.Run("group", func(t *testing.T) {
+		for _, testCase := range testCases {
+
+			testCase := testCase
+
+			t.Run(testCase.name, func(t *testing.T) {
+				t.Parallel()
+
+				publishMessage(ctx, client, testCase.topicName, testCase.expected)
+				actual := receiveMessage(ctx, client, t, testCase.subName)
+
+				assert.Equal(t, testCase.expected, actual)
```

Comment:
This test is no longer actually testing the message. It will *always* pass because the test is using the same variable for comparing expected and actual. 

For example, it will still pass even if you change the expected value to "I-can-put-any-value-here-and-my-test-will-still-pass". They should be 2 separate values, e.g. expected = "my-pub-sub-message" and actual = "my-pub-sub-message" and then call assert(expected, actual). The current setup is similar to assert (expected, expected)

Reply:
Could you elaborate more about using the same variable for assertion?

In my view, there are two variables: `testCase.expected` and `actual`.

`testCase.expected` is declared and assigned in the `testCase` struct while `actual` is declared and assigned when `receiveMessage` returns.

The test will pass when the message is correctly received regardless of what message is published. This means the topic and subscription pair are wired up correctly.

The test will fail when the message received (`actual`) is different from `testCase.expected` (functional issue) or the topic and subscription are not wired up properly (environment setup issue).